python-scripts/eeprom-util.py

```diff
@@ -0,0 +1,200 @@
+# object (i / j) which is necessary when we call eepKeyCrypt on the next chunk of data
+# from the Eeprom method object (eeprom.hddKey + eeprom.gameReg). This is also why we
+# don't use any standard RC4 crypto libaries here.
+
+# This code was adapted from https://en.wikipedia.org/wiki/RC4
+
+# store RC4 key permuataion data here
+class EepKeyObj:
+  state = bytearray(range(256))
+  i = 0
+  j = 0
+
+# RC4 Key-Scheduling Algorithm (KSA)
+# because of the confounder we only want to call this once
+def eepKeyPrepCrypt(eepKey, eepKeyObj):
+  j = 0
+
+  for i in range(256):
+    j = (j + eepKeyObj.state[i] + eepKey[i % len(eepKey)]) % 256
+    eepKeyObj.state[i], eepKeyObj.state[j] = eepKeyObj.state[j], eepKeyObj.state[i]
+
+# RC4 Pseudo-Random Generation Algorithm (PRGA)
+# because of the confounder this has been slighly modified to keep
+# i / j stored in EepKeyObj rather than setting them to 0 per decryption
+def eepKeyCrypt(eepCryptData, eepKeyObj):
+  eepData = bytearray()
+  i = eepKeyObj.i
+  j = eepKeyObj.j
+  xorInd = 0
+
+  for byte in eepCryptData:
+    i = (i + 1) % 256
+    j = (j + eepKeyObj.state[i]) % 256
+    eepKeyObj.state[i], eepKeyObj.state[j] = eepKeyObj.state[j], eepKeyObj.state[i]
+    eepData.append(byte ^ eepKeyObj.state[(eepKeyObj.state[i] + eepKeyObj.state[j]) % 256])
+
+  eepKeyObj.i = i
+  eepKeyObj.j = j
+  return eepData
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-f', action='store', dest='eepFileLoc',
+                  help='location of your Xbox eeprom dump')
+parser.add_argument('-k', action='store', dest='keyFileLoc',
+                  help='location of your Xbox eeprom key dump')
+args = parser.parse_args()
+
+if args.eepFileLoc:
+  eepFile = open(args.eepFileLoc, 'rb')
+  eepDump = eepFile.read(256)
+  eepFile.close()
+  eeprom = Eeprom(eepDump)
+  print ('ENCRYPTED EEPROM DATA')
+  print ('EEPROM Hash:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.hash)))
+  print ('Encrypted Confounder:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.confounder)))
+  print ('Encrypted HDD Key:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.hddKey)))
+  print ('Encrypted Game Region Code:\t{}'.format(''.join(format(x, '02x') for x in eeprom.gameReg)))
+
+  print ('\nIDENTITY EEPROM DATA')
+  print ('Identity Checksum:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.identChk)))
+  print ('Serial Number:\t\t\t{}'.format(eeprom.serial.decode('ascii')))
+  print ('MAC Address:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.mac)))
+  print ('Online Key:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.onlineKey)))
+
+  # print NTSC / PAL instead of their respective codes
+  if int.from_bytes(eeprom.videoStd, 'big', signed=False) == 81920: # 81920 == 00014000 
+    print ('Video Standard:\t\t\tNTSC ({})'.format(''.join(format(x, '02x') for x in eeprom.videoStd)))  
+  elif int.from_bytes(eeprom.videoStd, 'big', signed=False) == 229376: # 229376 == 00038000
```

Comment:
nit: please use hex constants

Reply:
No, little endian will reverse the byte order, this is stored in memory already in the correct order

Reply:
To use hex contents we'd have to convert to string as hex() expects a string in python. We could use byte notation, but that will look really ugly. It used to be string, but i changed it due to a previous complaint.

Reply:
```
[fox@x230 ~]$ python3
Python 3.6.2 (default, Jul 20 2017, 03:52:27) 
[GCC 7.1.1 20170630] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print(0x100)
256
```